pkg/cloudprovider/providers/openstack/openstack.go

```diff
@@ -555,8 +557,60 @@ func (apiVersions APIVersionsByID) Less(i, j int) bool {
 	return apiVersions[i].ID > apiVersions[j].ID
 }
 
-func (os *OpenStack) volumeService() (volumeService, error) {
-	switch os.bsOpts.BSVersion {
+func autoVersionSelector(apiVersion *apiversions_v1.APIVersion) string {
+	switch apiVersion.ID {
+	case "v2.0":
+		return "v2"
+	case "v1.0":
+		return "v1"
+	default:
+		return ""
+	}
+}
+
+func doBsApiVersionAutodetect(availableApiVersions []apiversions_v1.APIVersion, AllowAutoSupported bool) string {
+	// search for CURRENT api version
+	// There should be only one api version with CURRENT status
+	for _, version := range availableApiVersions {
+		if version.Status == "CURRENT" {
+			if detectedApiVersion := autoVersionSelector(&version); detectedApiVersion != "" {
+				glog.V(3).Infof("Blockstorage API version probing has found a suitable CURRENT api version: %s", detectedApiVersion)
+				return detectedApiVersion
+			} else {
+				break
+			}
+		}
+	}
+	// Only if we don't support CURRENT api version we get this far
+	glog.Warningf("I don't support the CURRENT blockstorage API version")
+	if AllowAutoSupported {
+		glog.Info("Blockstorage api version autodetection is allowed to search in SUPPORTED versions")
+		sort.Sort(APIVersionsByID(availableApiVersions))
+		for _, version := range availableApiVersions {
+			if version.Status == "SUPPORTED" {
+				if detectedApiVersion := autoVersionSelector(&version); detectedApiVersion != "" {
+					glog.V(3).Infof("Blockstorage API version probing has found a suitable SUPPORTED api version: %s", detectedApiVersion)
+					return detectedApiVersion
+				}
+			}
+		}
+	}
+	// Nothing suitable found, failed autodetection
+	errMsg := "Autoprobing for blockstorage API failed to find a suitable api version See auto-allow-supported and bs-version settings."
```

Comment:
Moved and print settings. But really auto-allow-supported may change bs-version will always be auto